app/routines/create_user.rb

```diff
@@ -41,19 +29,33 @@ def exec(username:, title: nil, first_name: nil, last_name: nil,
     transfer_errors_from(outputs[:user], {type: :verbatim})
   end
 
-  # Returns a valid, though possibly non-unique, username
-  def get_valid_username(base)
-    base = randomify_username(base) if base.blank?
+  def generate_unique_valid_username(username)
+    return username if username_is_valid?(username) && username_is_unique?(username)
 
-    # Go ahead and sanitize the username knowing that User.create will
-    # expect it to adhere to certain rules on length and content.
-    base = base.gsub(User::USERNAME_DISCARDED_CHAR_REGEX, '')
-               .slice(0..User::USERNAME_MAX_LENGTH - 1).downcase
+    loop do
+      username = create_random_username
+      break if username_is_unique?(username)
+    end
```

Comment:
The loop breaks out of it on our terms—only once we create a unique username using rand(1000000). The loop breaking out is not dependent on any user input. Correct me if I'm wrong.